pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -0,0 +1,539 @@
+		interval:         interval,
+		client:           client,
+		networkByFamily:  networkByFamily,
+		serviceLister:    serviceInformer.Lister(),
+		servicesSynced:   serviceInformer.Informer().HasSynced,
+		ipAddressLister:  ipAddressInformer.Lister(),
+		ipAddressSynced:  ipAddressInformer.Informer().HasSynced,
+		svcQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "services"),
+		ipQueue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ipaddresses"),
+		workerLoopPeriod: time.Second,
+		broadcaster:      eventBroadcaster,
+		recorder:         recorder,
+		clock:            clock.RealClock{},
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				r.svcQueue.Add(key)
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				r.svcQueue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			// IndexerInformer uses a delta queue, therefore for deletes we have to use this
+			// key function.
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				r.svcQueue.Add(key)
+			}
+		},
+	}, interval)
+
+	ipAddressInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				r.ipQueue.Add(key)
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				r.ipQueue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			// IndexerInformer uses a delta queue, therefore for deletes we have to use this
+			// key function.
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				r.ipQueue.Add(key)
+			}
+		},
+	}, interval)
+
+	return r
+}
+
+// RunUntil starts the controller until the provided ch is closed.
+func (r *RepairIPAddress) RunUntil(onFirstSuccess func(), stopCh chan struct{}) {
+	defer r.ipQueue.ShutDown()
+	defer r.svcQueue.ShutDown()
+	r.broadcaster.StartRecordingToSink(stopCh)
+	defer r.broadcaster.Shutdown()
+
+	klog.Info("Starting ipallocator-repair-controller")
+	defer klog.Info("Shutting down ipallocator-repair-controller")
+
+	if !cache.WaitForNamedCacheSync("ipallocator-repair-controller", stopCh, r.ipAddressSynced, r.servicesSynced) {
+		return
+	}
+
+	// First sync goes through all the Services and IPAddresses in the cache
+	// once syncer, it signals the main loop and works using the handlers, since
+	// it's less expensive and more optimal.
+	if err := r.runOnce(); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	onFirstSuccess()
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(r.ipWorker, r.workerLoopPeriod, stopCh)
```

Comment:
Have we considered the ramification of having two loops operating on the same object graph? wouldn't it make more sense (and less racy) to work on services then IPs?

Reply:
The two loops are completely independent, Services loop: check that each clusterIP has an associated IPAddress and that this IPAddress is correct, if it doesn't exist or is wrong it repairs.

IPAddress loop: check that each IPAddress is associated to the corresponding Service, if the IP is orphan it deletes it, but it can only be orphan if the service is referencing doesn't exist

The added complexity has a great benefit that is performance, instead of working every X mins for doing the reconciliation, every event is processed as soons is received ... there is an integration test that goes from 3min to repair an IPAddress to ~5 sec with this